In [1]:
import pandas as pd
import numpy as np 
import os 

# Importing Data Dictionaries

In [2]:
#importing data dictionaries 
path=(r"C:\Users\hardjan\Instacart Basket Analysis March 2024")
#Importing products data set 
df_prods= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
# Importing orders data set 
df_ords=pd.read_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [3]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_sequence_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06           3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05           1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04           1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00           1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04           5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05           1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05           2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05           1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [4]:
# create a data frame 
df_test=pd.DataFrame()


In [5]:
# create a mix type column 
df_test['mix']=['a','b',1,True]

In [6]:
df_test

mix
0     a
1     b
2     1
3  True

In [7]:
# check for mixed types (instead of applymap here we use the data frame .map)
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# second way of finding mixed types 

In [8]:
for col in df_ords.columns .tolist():
    if len (df_ords[col].apply(type).unique())>1:
        print (col+ 'have mix data type')
    else:
        print(col + 'dont have mix data type')

Unnamed: 0dont have mix data type
order_iddont have mix data type
user_iddont have mix data type
order_sequence_numberdont have mix data type
orders_day_of_weekdont have mix data type
order_hour_of_daydont have mix data type
days_since_prior_orderdont have mix data type


# finding null values

In [9]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
# finding all the 16 values in the above output 
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# different ways to address missing values

In [12]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [13]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
#drop missing values 
df_prods.dropna(inplace=True)


In [16]:
# drop only the NaNs from a particular column 
df_prods.dropna(subset=['product_id'], inplace= True)

In [17]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

# finding duplicate values 

In [18]:
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
df_prods_clean.drop_duplicates()

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [21]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape



(49672, 5)

# Task Workings 4.5

In [23]:
# 02. running describe for the orders data frame 
df_ords.describe()

Unnamed: 0      order_id       user_id  order_sequence_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06           3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05           1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04           1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00           1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04           5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05           1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05           2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05           1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#There are many missing values in the 'days_since_prior_order' column, as its count is lower than the other columns´.
'order_number' has min value as 1 and max value as 100. Q2 or median is 11. No discrepancies found.
'days_since_prior_order' has min value as 0 and max as 30. Q2 or median is 7. No discrepancies found.


In [24]:
# 3 Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
    if len(df_ords[col].apply(type).unique())>1:
        print(col+' have mix datatype')
    else:
        print(col+' don\'t have mix datatype')

Unnamed: 0 don't have mix datatype
order_id don't have mix datatype
user_id don't have mix datatype
order_sequence_number don't have mix datatype
orders_day_of_week don't have mix datatype
order_hour_of_day don't have mix datatype
days_since_prior_order don't have mix datatype


# none of the columns have mixed data types

In [25]:
# 4 Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()


Unnamed: 0                     0
order_id                       0
user_id                        0
order_sequence_number          0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

 There are 206,209 missing values in the 'days_since_prior_order' column. These are probably first orders placed by customers, therefore any order placed for the first time will have a missing value in the 'days_since_prior_order' column.

Removing the missing values would result in the loss of important information about customers who just made their first order, therefore here are the steps of how I plan to address the issue
 
1. Checking the frequency of missing values is the same as the frequency of ordesequence_r_number=1 .
2. Create a new column that flags new orders (True or False).

In [26]:
# 6 Address the missing values using an appropriate method
# Create a subset containing only missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [27]:
df_nan

Unnamed: 0  order_id  user_id  order_sequence_number  \
0                 0   2539329        1                      1   
11               11   2168274        2                      1   
26               26   1374495        3                      1   
39               39   3343014        4                      1   
45               45   2717275        5                      1   
...             ...       ...      ...                    ...   
3420930     3420930    969311   206205                      1   
3420934     3420934   3189322   206206                      1   
3421002     3421002   2166133   206207                      1   
3421019     3421019   2227043   206208                      1   
3421069     3421069   3154581   206209                      1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 7 columns]

# Adressing the missing values by creating a new column  named 'new_customer'

In [28]:
df_ords_clean=df_ords
df_ords_clean['new_customer']=df_ords['days_since_prior_order'].isnull ()==True
df_ords_clean

Unnamed: 0  order_id  user_id  order_sequence_number  \
0                 0   2539329        1                      1   
1                 1   2398795        1                      2   
2                 2    473747        1                      3   
3                 3   2254736        1                      4   
4                 4    431534        1                      5   
...             ...       ...      ...                    ...   
3421078     3421078   2266710   206209                     10   
3421079     3421079   1854736   206209                     11   
3421080     3421080    626363   206209                     12   
3421081     3421081   2977660   206209                     13   
3421082     3421082    272231   206209                     14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                         2                  8                     NaN   
1                         3                  7                    15.0   
2                         3                 12                    21.0   
3                         4                  7                    29.0   
4                         4                 15                    28.0   
...                     ...                ...                     ...   
3421078                   5                 18                    29.0   
3421079                   4                 10                    30.0   
3421080                   1                 12                    18.0   
3421081                   1                 12                     7.0   
3421082                   6                 14                    30.0   

         new_customer  
0                True  
1               False  
2               False  
3               False  
4               False  
...               ...  
3421078         False  
3421079         False  
3421080         False  
3421081         False  
3421082         False  

[3421083 rows x 8 columns]

In [29]:
# Check the frequency of order_number=1 in df_nan
df_nan['order_sequence_number'].value_counts()

order_sequence_number
1    206209
Name: count, dtype: int64

In [30]:
# Check the frequency of order_number=1 in df_ords
df_ords['order_sequence_number'].value_counts()

order_sequence_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

## The frequency of missing values (206,209) is the same as the frequency of order_number=1.

## Missing values handled correctly 


In [31]:
# 7 Run a check for duplicate values in your df_ords data
df_ords_dups=df_ords[df_ords.duplicated()]

In [32]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_sequence_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

In [33]:
df_ords_dups.shape

(0, 8)

# No Duplicates found 

In [34]:
#8 exporting data frames 
# Exporting df_prods_clean_no_dups dataframe as "products_checked.csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [35]:
# Exporting df_ords dataframe as "orders_checked.csv"
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))